File: projects/grind-75/typescript/array/insert-interval.ts
export function insert(
  intervals: number[][],
  newInterval: number[],
): number[][] {
  let res: number[][] = [];

  for (let i = 0; i < intervals.length; i++) {
    const interval = intervals[i];
    if (interval[1] < newInterval[0]) {
      res.push(intervals[i]);
    } else if (interval[0] > newInterval[1]) {
      res.push(newInterval);
      return res.concat(intervals.slice(i));
    } else {
      newInterval = [
        Math.min(interval[0], newInterval[0]),
        Math.max(interval[1], newInterval[1]),
      ];
    }
  }
  return res.concat(newInterval);
}

File: projects/grind-75/python/array/insert_interval.py
from typing import List


class Solution:
    @classmethod
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        res = []

        for i, interval in enumerate(intervals):
            if newInterval[1] < interval[0]:
                res.append(newInterval)
                return res + intervals[i:]
            elif interval[1] < newInterval[0]:
                res.append(interval)
            else:
                newInterval = [
                    min(newInterval[0], interval[0]),
                    max(newInterval[1], interval[1]),
                ]

        res.append(newInterval)
        return res


"""
Time Complexity - O(n)
Space Complexity - O(n)
"""
